mods/fireflies/init.lua

```diff
@@ -0,0 +1,170 @@
+	description = "Firefly",
+	drawtype = "plantlike",
+	tiles = {{
+		name = "fireflies_firefly_animated.png",
+		animation = {
+			type = "vertical_frames",
+			aspect_w = 16,
+			aspect_h = 16,
+			length = 1.5
+		},
+	}},
+	inventory_image = "fireflies_firefly.png",
+	wield_image =  "fireflies_firefly.png",
+	waving = 1,
+	paramtype = "light",
+	sunlight_propagates = true,
+	buildable_to = true,
+	walkable = false,
+	groups = {catchable = 1},
+	selection_box = {
+		type = "fixed",
+		fixed = {-0.1, -0.1, -0.1, 0.1, 0.1, 0.1},
+	},
+	light_source = 6,
+	floodable = true,
+	on_flood = function(pos, oldnode, newnode)
+		minetest.add_item(pos, "fireflies:firefly 1")
+	end
+})
+
+
+-- bug net
+minetest.register_tool("fireflies:bug_net", {
+	description = "Bug Net",
+	inventory_image = "fireflies_bugnet.png",
+	on_use = function(itemstack, player, pointed_thing)
+		if not pointed_thing or pointed_thing.type ~= "node" or
+				minetest.is_protected(pointed_thing.under, player:get_player_name()) then
+			return
+		end
+		local node_name = minetest.get_node(pointed_thing.under).name
+		local inv = player:get_inventory()
+		if minetest.get_item_group(node_name, "catchable") == 1 then
+			minetest.set_node(pointed_thing.under, {name = "air"})
+			local stack = ItemStack(node_name.." 1")
+			local leftover = inv:add_item("main", stack)
+			if leftover:get_count() > 0 then
+				minetest.add_item(pointed_thing.under, node_name.." 1")
+			end
+		end
+		if not minetest.setting_getbool("creative_mode") then
+			itemstack:add_wear(256)
+			return itemstack
+		end
+	end
+})
+
+minetest.register_craft( {
+	output = "fireflies:bug_net",
+	recipe = {
+		{"farming:string", "farming:string", ""},
+		{"farming:string", "farming:string", ""},
+		{"default:stick", "", ""}
+	}
+})
+
+
+-- firefly in a bottle
+minetest.register_node("fireflies:firefly_bottle", {
+	description = "Firefly in a Bottle",
+	inventory_image = "fireflies_bottle.png",
+	wield_image = "fireflies_bottle.png",
+	tiles = {{
+		name = "fireflies_bottle_animated.png",
+		animation = {
+			type = "vertical_frames",
+			aspect_w = 16,
+			aspect_h = 16,
+			length = 1.5
+		},
+	}},
+	drawtype = "plantlike",
+	paramtype = "light",
+	sunlight_propagates = true,
+	light_source = 9,
+	walkable = false,
+	groups = {snappy = 3, attached_node = 1},
+	selection_box = {
+		type = "fixed",
+		fixed = {-0.25, -0.5, -0.25, 0.25, 0.3, 0.25}
+	},
+	sounds = default.node_sound_glass_defaults(),
+	on_rightclick = function(pos, node, player, itemstack, pointed_thing)
+		local lower_pos = {x = pos.x, y = pos.y + 1, z = pos.z}
+		if minetest.is_protected(pos, player:get_player_name()) or
+				minetest.get_node(lower_pos).name ~= "air" then
+			return
+		end
+
+		local upper_pos = {x = pos.x, y = pos.y + 2, z = pos.z}
+		local firefly_pos
+
+		if minetest.is_protected(upper_pos, player:get_player_name()) or
+				minetest.get_node(upper_pos).name ~= "air" then
+			firefly_pos = lower_pos
```

Comment:
The lower pos is not checked for protection if the firefly is placed there. Sorry maybe we miscommunicated.

The bottle should be checked for protection, and the pos chosen for firefly placement should be checked for protection, so 2 checks.
So the code has to first decide what pos is used for placement, according to available air nodes, then check that pos for protection.

Reply:
That doesn't make sense.

Let's say lower_pos is air and unprotected, and upper_pos is air and protected, your approach would lead to the firefly not being released, which is incorrect. 

I'd like to restore my previous method. The additional check is needed to do this properly, and I don't want the checks incorrectly handled just to save on one check.

Reply:
I think we misunderstand each other.

> Let's say lower_pos is air and unprotected, and upper_pos is air and protected, your approach would lead to the firefly not being released, which is incorrect.

Not sure.

If the firefly is going to be placed in upper (because both nodes are air, the most common situation), protection will of course need to be checked for bottle and upper, it's highly unlikely there will be a 1-node-high protected volume between them, even if there is it doesn't matter because nothing is being altered there, it's just protected air which remains air.
So in this situation only 2 protection checks are needed for bottle and upper.
However the first thing the function does is check protection for lower.

If upper turns out to be protected air, then check protection for lower, in this case (less common case) all 3 checks are needed: bottle, upper and lower.

So it seems to me the logic is:

Check protection on bottle, if protected, return.
Get lower and upper nodes.
If both air, check protection for upper, if unprotected, place.
Else if lower is air, check protection for lower, if unprotected, place.
Else return.

I may be wrong.

Reply:
However if that doesn't make sense to you, feel free to return to your previous code as it's not a big issue.

Reply:
Okay, so not discarding the third check but limiting it.  That's simple enough, it's just there is no way to remove it completely.

Reply:
Yes that's where i went wrong, in one case 3 checks are needed.

Reply:
Feels messier to me, but done.